Make templates extensible and template context configurable by other extensions

fix-fix opened the pull request:
Current template configuration allows specifying only a single directory from which you can reference other templates or partials, so it's not possible to modify document template in a composable way (by changing only the things you need and keeping everything else intact).

Another problem is that it's not possible to adjust some parts of the document without fully copying it into a separate template. While there is a powerful system for adding markup to the <head> through react-helmet from `settings.runtime.head`, some external scripts and other resources should be included in the <body> tag to prevent main content loading and rendering delay.

To fix this problem the following changes are proposed:
* add ability to specify multiple templates lookup paths. Nunjucks provides ability to add multiple paths for templates lookup in [`configure()`](https://mozilla.github.io/nunjucks/api.html#configure), so it makes sense to allow not only a single path, but an array of paths too. Name of the default template was changed to prevent name conflicts.
* add ability to provide template context from extensions. Right now some data can be passed in one form or another, but it's either limited (only some parts of Roc config are passed), exposed to the client, or not suitable for runtime generation (like specifying a file in `build.templateValues`).

This change allows extensions to extend template directories list and pass custom data to them, which opens some new possibilities, like:
* creating an extensible base template in custom extensions: the package will only have to add its templates path to the list
* offloading non-essential to the project information (like markup for analytics tracking, performance measurements, etc.), without copying them, but at the same time making it configurable

It's possible to also modify default template `roc-base-template.njk` and group markup into blocks for easy `{% extends %}` without creating a separate template, but I think it will add too much noise and still won't suit everyone's needs

---

For example one can create their own extensible template _my-base-template.njk_ and add it to the paths:

```nunjucks
<html>
  ...
  <body>
    ...
    {% block bodyScripts %}
    <script src="{{ bundleName }}"></script>
    {% endblock %}
```

And then in their project template _index.njk_:
```nunjucks
    {% extends "my-base-template.njk" %}
    {% block bodyScripts %}
    {{ super() }}
    <script src="https://third.party/script.js"></script>
    <noscript><iframe src="https://third.party/script-fallback.html" height="0" width="0" style="display:none;visibility:hidden"></iframe></noscript>
    {% endblock %}
```

dlmr commented on extensions/roc-package-web-app-react/src/app/server/reactRenderer.js:
```diff
@@ -28,8 +29,11 @@ const whiteListed = () => (
-    const templatePath = rocConfig.runtime.template.path || `${myPath}/views`;
-    nunjucks.configure(getAbsolutePath(templatePath), {
+    const templatePaths = [defaultTemplatePath]
+        .concat(rocConfig.runtime.template.path || [])
```
Should we change the order here, with the user specified directory coming before the default one? That would mean that we could keep the template name while still allowing this behavior.

dlmr commented on extensions/roc-package-web-app-react/src/app/server/reactRenderer.js:
```diff
@@ -58,6 +62,7 @@ export function initRenderPage({ script, css }, distMode, devMode, Header) {
         }
 
         return nunjucks.render(rocConfig.runtime.template.name, {
+            ...build.templateContext,
```
This seems not to be documented in the settings, is it missing?

dlmr commented on extensions/roc-package-web-app-react/src/config/roc.config.js:
```diff
@@ -26,7 +26,7 @@ export default {
             template: {
                 path: undefined,
-                name: 'main.html',
+                name: 'roc-base-template.njk',
```
I would like to keep the default template name for now since this otherwise would create breaking changes in a lot of projects that only have changed the template directory but not the template name.

dlmr commented:
Thanks for the PR!

In general I think we should try to keep the PRs focused and do updating of dependencies that are unrelated for the overall change to another PR. My suggestion is also do wait with regenerating the documentation until we do a new release, that way the newly generated documentation that we commit into Git becomes a sort of changelog of all the changes that has been done since the last release.

Another thought if maybe we need something other than an array in the settings for the template paths. This since it is not very composable since only one extension or project can define something without it being overridden. Maybe we could use an `action` for example that would allow for a more flexible structure.

> It's possible to also modify default template roc-base-template.njk and group markup into blocks for easy {% extends %} without creating a separate template, but I think it will add too much noise and still won't suit everyone's needs

A good thought and something we definitely should consider, but only if it creates enough value as you mention.

fix-fix commented:
> My suggestion is also do wait with regenerating the documentation until we do a new release

Yeah, totally agree, I've cherry-picked this commit from another branch and haven't noticed changes in markdown files even before submitting PR. I will revert that, thanks.

As for dependencies update I will check if they are still needed at all, and extract it to another PR.

> Maybe we could use an action for example that would allow for a more flexible structure.

I considered defining an action for templates, but decided to keep things simplier for the base case (i.e. overriding), because projects and extensions can still use 'update-settings' for defining their own. I will look into creating an action similar to other list values.

All other comments look reasonable too, I definitely want to keep things backward-compatible. I'll update the PR soon.

fix-fix commented:
New changes:
* previous HTML template (since most changes were in formatting and would pollute git history) and its settings were restored to keep existing projects working
* added docs for `build.templateContext` setting
* changed templates lookup paths priority
* added hook for adding new paths, similar to 'get-webpack-targets'

I also removed commits with deps and docs update and moved them to #69

dlmr commented:
Thanks again for this!

This is a great start that we can build on top of, making the template system even more composable in the future. 

One thing that I would like to change is to make it possible for extensions to define template values, here called `templateContext`, using a hook for this which would make composition a little easier and more powerful.

I also like what we have started here with `extends`. It should be possible to take this one step further; allowing extensions to extend each other in a chain for more possibilities.

Let's revisit this soon in another PR.
